proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lower_iMM`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions traps if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+```
+
+Currently, only UTF8 is supported as an encoding. UTF16 may be added in the future.
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding $alloc: #funcidx
+```
+
+The type of this instruction is:
+```
+[string] -> [i32, i32]
+
+Where:
+  * $alloc has a valid functype of [i32] -> [i32]
+```
+
+The instruction will execute the following steps:
+```
+  1. Let $length be the bytes required to encode the string with #encoding
+  2. Call $alloc with $length and receive an offset into #memidx
+  3. Encode the string into the offset at memory
+```
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
+
+The type of this instruction is:
+```
+[$a, $b, ..] -> [$type]
+
+Where:
+  * $type is a type of the form (record $fields) taken from #typeidx
+  * $a, $b, .. are the #interface-valtype's for each field in $fields
+```
+
+The output is a record interface value with fields given by the inputs.
+
+## `record.lower`
+
+The full syntax of this instruction is:
+```
+record.lower #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [$fields]
+
+Where:
+  * $type is a valid type of the form (record $fields) taken from #typeidx
+```
+
+The outputs are the fields of the input `record`.
+
+## `array.lift_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lift_memory #typeidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$base: i32, $length: i32] -> [$type]
+
+Where:
+  * $type is a type of the form (array $inner) taken from #typeidx
+  * #interface-instr* is valid with type [i32] -> [$inner]
+```
+
+This instruction will execute the following steps:
+```
+  1. Let $offsets = take([$base, $base + $width, $base + $width * 2, ..], $length)
+  2. Map #interface-instr* upon $offsets to yield the elements of the array
+```
+
+## `array.lower_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lower_memory #typeidx #memidx $alloc: #funcidx $width: #u32
+  interface-instr*
+end
+```
```

Comment:
dont need functions here. It makes reasoning about the adapter fusion process more difficult.

Reply:
How does that make reasoning about the fusion process difficult? I've used the pattern of performing memory allocation in instructions that lower interface-types to linear memory to match with the 'Explainer'.

Reply:
The real task for the fusion process is to move code around - arbitrary (-ish) bits of it. Forcing it to understand that some lambdas are intended to be unfolded just makes it harder.
I am quite concerned that adapters should be fusable with very limited reasoning required. One way of visualizing this is to imagine that you have to implement a fusion in 100 microseconds (which is roughly 10000 X64 instructions on high-end hardware). Even with that, it would take 10 seconds to produce the fusions for 100K adapters. There are already workloads with that many import/exports.